File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        Map<String, Integer> index = new HashMap<>();
        int[][] graph = new int[friends.length][friends.length];
        int[] factor = new int[friends.length];
        int answer = 0;
        
        for(int i = 0; i < friends.length; i++) index.put(friends[i], i);
        for(int i = 0; i < gifts.length; i++) {
            String[] tmp = gifts[i].split(" ");
            factor[index.get(tmp[0])]++;
            factor[index.get(tmp[1])]--;
            graph[index.get(tmp[0])][index.get(tmp[1])]++;
        }
        
        for(int i = 0; i < graph.length; i++) {
            int max = 0;
            for(int j = 0; j < graph.length; j++) {
                if(i == j) continue;
                if(graph[i][j] > graph[j][i]) max++;
                else if (graph[i][j] == graph[j][i] && factor[i] > factor[j]) max++;
            }
            answer = Math.max(answer, max);
        }
        
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    N = len(friends)
    I = {name : index for name, index in zip(friends, range(N))}
    
    record = [[0] * N for _ in range(N)]
    give = [0] * N
    recive = [0] * N
    
    for gift in gifts:
        g, r = gift.split()
        give[I[g]] += 1
        recive[I[r]] += 1
        record[I[g]][I[r]] += 1
        
    score = [0] * N
    for x in range(N):
        score[x] = give[x] - recive[x]
    
    answer = [0] * N
    for i in range(N):
        for j in range(N):
            if i == j: continue
            case1 = (record[i][j] > record[j][i])
            case2 = (record[i][j] == record[j][i] and score[i] > score[j])
            answer[i] += (case1 or case2)
    
    return max(answer)